### ISSUE TITLE
False positive in TypeName check for numeric underscores (Google style §5.3)

### ISSUE THREAD

**Issue by VeniceWish** [checkstyle/checkstyle#17716]
I have read check documentation: https://checkstyle.org/checks/naming/typename.html#TypeName
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

From: https://google.github.io/styleguide/javaguide.html#s5.3-camel-case

> In very rare circumstances (for example, multipart version numbers), you may need to use underscores to separate adjacent numbers, since numbers do not have upper and lower case variants.

Code:
```java
class Test {
  class MüllersAlgorithm {} // Expected Violation

  class Guava33_4_6 {} // Should be allowed by Google style
}

```

Cli:
```
$ java -jar checkstyle-10.26.1-all.jar -c google_checks.xml Test.java 
Starting audit...
[WARN] Test.java:2:9: Type name 'MüllersAlgorithm' must match pattern '^[A-Z][a-zA-Z0-9]*$'. [TypeName]
[WARN] Test.java:4:9: Type name 'Guava33_4_6' must match pattern '^[A-Z][a-zA-Z0-9]*$'. [TypeName]
Audit done.
```

---
According to the [Google Java Style Guide, §5.3 Camel case](https://google.github.io/styleguide/javaguide.html?utm_source=chatgpt.com#s5.3-camel-case), “In very rare circumstances (for example, multipart version numbers), you may need to use underscores to separate adjacent numbers, since numbers do not have upper and lower case variants.”

In the example below, a type name with underscores between numeric sequences is valid per Google style, but Checkstyle with google_checks.xml reports a violation:
```java
class Guava33_4_6 {}  // Should be allowed
```
This appears to be a configuration issue in Checkstyle’s Google style, where
```xml
<property name="format" value="^[A-Z][a-zA-Z0-9]*$"/>
```
is currently set, but should allow underscores only between numbers to align with Google Java Style. For example:
```xml
<property name="format"
          value="^[A-Z][A-Za-z0-9]*$|^[A-Z](?:[A-Za-z0-9]*\d+(?:_\d+)+[A-Za-z0-9]*)*$"/>

```
Apart from `TypeName`, other checks that enforce upper camel case or lower camel case naming may also need similar adjustments.

ATTENTION: revert this update https://github.com/checkstyle/checkstyle/pull/18442/changes#r2659944571 to restored coverage to full.

**Comment by romani** [comment]
There is some activity on this problem at https://github.com/checkstyle/checkstyle/pull/17708

But we will keep this issue open, to make sure we cover all cases.

**Comment by VeniceWish** [comment]
Thanks!

**Comment by vivek-0509** [comment]
@romani the issue is already fixed, we can close this

**Comment by romani** [comment]
@vivek-0509 , please share CLI output over latest release and issue author code to be confident.

**Comment by vivek-0509** [comment]
@romanii here is the CLI ouput issue is fixed.

```
vivek@Viveks-MacBook-Air checkstyle % cat > /tmp/checkstyle-test/Test.java << 'EOF'
class Test {
  class MüllersAlgorithm {} // Expected Violation

  class Guava33_4_6 {} // Should be allowed by Google style
}
EOF

vivek@Viveks-MacBook-Air checkstyle % cp /Users/vivek/checkstyle/src/main/resources/google_checks.xml /tmp/checkstyle-test/

vivek@Viveks-MacBook-Air checkstyle % java -Duser.language=en -Duser.country=US -jar /Users/vivek/checkstyle/target/checkstyle-*-all.jar -c /tmp/checkstyle-test/google_checks.xml /tmp/checkstyle-test/Test.java
Starting audit...
[WARN] /tmp/checkstyle-test/Test.java:2:9: Type name 'MüllersAlgorithm' must match pattern '^[A-Z][a-zA-Z0-9]*(?:[0-9](?:_[0-9]+)*)?$'. [TypeName]
Audit done.
```

**Comment by romani** [comment]
in future please use `java -Duser.language=en -Duser.country=US -jar /Users/vivek/checkstyle/target/checkstyle-*-all.jar -c /google_checks.xml /tmp/checkstyle-test/Test.java`

attention to `/google_checks.xml`  it is usage of google style from jar , very proven way to show that all is released , no nuances in misconfiguration
